packages/compiler/emitter-framework/custom-key-map.ts

```diff
@@ -0,0 +1,53 @@
+/**
+ * This is a map type that allows providing a custom keyer function. The keyer
+ * function returns a string that is used to look up in the map. This is useful
+ * for implementing maps that look up based on an arbitrary number of keys.
+ *
+ * For example, to look up in a map with a [ObjA, ObjB)] tuple, such that tuples
+ * with identical values (but not necessarily identical tuples!) create an
+ * object keyer for each of the objects:
+ *
+ *     const aKeyer = CustomKeyMap.objectKeyer();
+ *     const bKeyer = CUstomKeyMap.objectKeyer();
+ *
+ * And compose these into a tuple keyer to use when instantiating the custom key
+ * map:
+ *
+ *     const tupleKeyer = ([a, b]) => `${aKeyer.getKey(a)}-${bKeyer.getKey(b)}`;
+ *     const map = new CustomKeyMap(tupleKeyer);
+ *
+ */
+export class CustomKeyMap<K extends readonly any[], V> {
```

Comment:
Can this/should this replace MultiKeyMap or should MultiKeyMap be built on top of this?

Reply:
It could, I suppose, but the interface is quite a bit different. Objections to leaving this for later?

Reply:
No objections, it seemed there was *some* overlap and I was mostly curious.